src/Controller/HtmlFormController.php

```diff
@@ -0,0 +1,31 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Retrofit\Drupal\Controller;
+
+use Drupal\Core\Controller\HtmlFormController as CoreHtmlFormController;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Retrofit\Drupal\Form\ArrayAccessFormState;
+use Symfony\Component\HttpFoundation\Request;
+
+class HtmlFormController extends CoreHtmlFormController
+{
+  /**
+   * @return mixed[]
+   */
+    public function getContentResult(Request $request, RouteMatchInterface $route_match): array
+    {
+        $form_arg = $this->getFormArgument($route_match);
+        $form_object = $this->getFormObject($route_match, $form_arg);
+        $form_state = new ArrayAccessFormState();
+        $request->attributes->set('form', []);
+        $request->attributes->set('form_state', $form_state);
+        $args = $this->argumentResolver->getArguments($request, [$form_object, 'buildForm']);
+        $request->attributes->remove('form');
+        $request->attributes->remove('form_state');
+        unset($args[0], $args[1]);
+        $form_state->addBuildInfo('args', array_values($args));
+        return $this->formBuilder->buildForm($form_object, $form_state);
+    }
```

Comment:
This recreates a lot of logic. My first thought is: can't we decorate the form builder to handle the form state in `\Drupal\Core\Form\FormBuilderInterface::buildForm`.